File: app/models/sections/ExamMaterial.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package models.sections;

import jakarta.persistence.Entity;
import jakarta.persistence.ManyToMany;
import java.util.Set;
import models.User;
import models.base.OwnedModel;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;
import org.springframework.beans.BeanUtils;

@Entity
public final class ExamMaterial extends OwnedModel {

    private String name;
    private String isbn;
    private String author;

    @ManyToMany(mappedBy = "examMaterials")
    private Set<ExamSection> examSections;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getIsbn() {
        return isbn;
    }

    public void setIsbn(String isbn) {
        this.isbn = isbn;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public Set<ExamSection> getExamSections() {
        return examSections;
    }

    public void setExamSections(Set<ExamSection> examSections) {
        this.examSections = examSections;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;

        if (!(o instanceof ExamMaterial that)) return false;

        return new EqualsBuilder().append(id, that.id).isEquals();
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 37).append(id).toHashCode();
    }

    public ExamMaterial copy(User user) {
        ExamMaterial material = new ExamMaterial();
        BeanUtils.copyProperties(this, material, "id", "examSections", "creator", "modifier");
        material.setCreatorWithDate(user);
        material.setModifierWithDate(user);
        return material;
    }
}

File: app/models/sections/ExamMaterial.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package models.sections

import jakarta.persistence.{Entity, ManyToMany}
import models.base.OwnedModel
import models.user.User

import scala.compiletime.uninitialized

@Entity
final class ExamMaterial extends OwnedModel:
  @ManyToMany(mappedBy = "examMaterials")
  var examSections: java.util.Set[ExamSection] = uninitialized

  var name: String   = uninitialized
  var isbn: String   = uninitialized
  var author: String = uninitialized

  def copy(user: User): ExamMaterial =
    val m = new ExamMaterial
    m.name = name
    m.isbn = isbn
    m.author = author
    m.setCreatorWithDate(user)
    m.setModifierWithDate(user)
    m

  override def equals(o: Any): Boolean = o match
    case e: ExamMaterial => this.id == e.id
    case _               => false

  override def hashCode: Int = id.toInt
